src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -14,211 +24,3020 @@ namespace Microsoft.CodeAnalysis.CSharp
+            }
+        }
+
+        /// <summary>
+        /// Used to resolve break statements in each statement form that has a break statement
+        /// (loops, switch).
+        /// </summary>
+        private void ResolveBreaks(TLocalState breakState, LabelSymbol label)
+        {
+            var pendingBranches = PendingBranches;
+            var count = pendingBranches.Count;
+
+            if (count != 0)
+            {
+                int stillPending = 0;
+                for (int i = 0; i < count; i++)
+                {
+                    var pending = pendingBranches[i];
+                    if (pending.Label == label)
+                    {
+                        Join(ref breakState, ref pending.State);
+                    }
+                    else
+                    {
+                        if (stillPending != i)
+                        {
+                            pendingBranches[stillPending] = pending;
+                        }
+                        stillPending++;
+                    }
+                }
+
+                pendingBranches.Clip(stillPending);
+            }
+
+            SetState(breakState);
+        }
+
+        /// <summary>
+        /// Used to resolve continue statements in each statement form that supports it.
+        /// </summary>
+        private void ResolveContinues(LabelSymbol continueLabel)
+        {
+            var pendingBranches = PendingBranches;
+            var count = pendingBranches.Count;
+
+            if (count != 0)
+            {
+                int stillPending = 0;
+                for (int i = 0; i < count; i++)
+                {
+                    var pending = pendingBranches[i];
+                    if (pending.Label == continueLabel)
+                    {
+                        // Technically, nothing in the language specification depends on the state
+                        // at the continue label, so we could just discard them instead of merging
+                        // the states. In fact, we need not have added continue statements to the
+                        // pending jump queue in the first place if we were interested solely in the
+                        // flow analysis.  However, region analysis (in support of extract method)
+                        // and other forms of more precise analysis
+                        // depend on continue statements appearing in the pending branch queue, so
+                        // we process them from the queue here.
+                        Join(ref this.State, ref pending.State);
+                    }
+                    else
+                    {
+                        if (stillPending != i)
+                        {
+                            pendingBranches[stillPending] = pending;
+                        }
+                        stillPending++;
+                    }
+                }
+
+                pendingBranches.Clip(stillPending);
+            }
```

Comment:
1. why does ResolveBreaks call SetState at the end, but this doesn't need to?
2. there looks to be a lot of common code patterns about walking these pending branches between ResolveBreaks/ResolveContinues/ResolveBranches.  Can we share anything here?

Reply:
Pausing here as it's 1am :)

Reply:
1. why does ResolveBreaks call SetState at the end, but this doesn't need to? Because the continue state is already the current state when this is called, and the change of state is processed on line 729.
2. there looks to be a lot of common code patterns about walking these pending branches between ResolveBreaks/ResolveContinues/ResolveBranches.  Can we share anything here? I would expect such refactoring to be outside the scope of this PR that just moves code from one file to another.


---
In reply to: [236573250](https://github.com/dotnet/roslyn/pull/31389#discussion_r236573250) [](ancestors = 236573250)